hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/SaslUtil.java

```diff
@@ -0,0 +1,195 @@
+              HADOOP_SECURITY_CRYPTO_CIPHER_KEY_BITLENGTH_KEY,
+              HADOOP_SECURITY_CRYPTO_CIPHER_KEY_BITLENGTH_DEFAULT) / 8;
+          CryptoCodec codec = CryptoCodec.getInstance(conf, suite);
+          byte[] inKey = new byte[keyLen];
+          byte[] inIv = new byte[suite.getAlgorithmBlockSize()];
+          byte[] outKey = new byte[keyLen];
+          byte[] outIv = new byte[suite.getAlgorithmBlockSize()];
+          assert codec != null;
+          codec.generateSecureRandom(inKey);
+          codec.generateSecureRandom(inIv);
+          codec.generateSecureRandom(outKey);
+          codec.generateSecureRandom(outIv);
+          return new CipherOption(suite, inKey, inIv, outKey, outIv);
+        }
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Encrypt the key and iv of the negotiated cipher option.
+   *
+   * @param option negotiated cipher option
+   * @param saslServer SASL server
+   * @return CipherOption negotiated cipher option which contains the
+   * encrypted key and iv
+   * @throws IOException for any error
+   */
+  public static CipherOption wrap(CipherOption option, SaslServer saslServer)
+      throws IOException {
+    if (option != null) {
+      byte[] inKey = option.getInKey();
+      if (inKey != null) {
+        inKey = saslServer.wrap(inKey, 0, inKey.length);
+      }
+      byte[] outKey = option.getOutKey();
+      if (outKey != null) {
+        outKey = saslServer.wrap(outKey, 0, outKey.length);
+      }
+      return new CipherOption(option.getCipherSuite(), inKey, option.getInIv(),
+          outKey, option.getOutIv());
+    }
+
+    return null;
+  }
+
+  /**
+   * Decrypt the key and iv of the negotiated cipher option.
+   *
+   * @param option negotiated cipher option
+   * @param saslClient SASL client
+   * @return CipherOption negotiated cipher option which contains the
+   * decrypted key and iv
+   * @throws IOException for any error
+   */
+  public static CipherOption unwrap(CipherOption option, SaslClient saslClient)
+      throws IOException {
+    if (option != null) {
+      byte[] inKey = option.getInKey();
+      if (inKey != null) {
+        inKey = saslClient.unwrap(inKey, 0, inKey.length);
+      }
+      byte[] outKey = option.getOutKey();
+      if (outKey != null) {
+        outKey = saslClient.unwrap(outKey, 0, outKey.length);
+      }
+      return new CipherOption(option.getCipherSuite(), inKey, option.getInIv(),
+          outKey, option.getOutIv());
+    }
+
+    return null;
+  }
+
+  public static List<CipherOption> getCipherOptions(Configuration conf)
+      throws IOException {
+    List<CipherOption> cipherOptions = null;
+    String cipherSuites = conf.get(HADOOP_SECURITY_CRYPTO_CIPHER_SUITE_KEY);
+    // Negotiate cipher suites if configured.  Currently, the only supported
+    // cipher suite is AES/CTR/NoPadding, but the protocol allows multiple
+    // values for future expansion.
+    if (cipherSuites != null && !cipherSuites.isEmpty()) {
```

Comment:
if (cipherSuites == null || cipherSuites.isEmpty())
  return null;
Reduce the level of nested.